### PR TITLE
Validation refactoring

### PR DESCRIPTION
**Pull request by vanruch**

### CODE REVIEW

File: src/services/entity_builder.js [hunk 1]
```diff
@@ -138,61 +136,40 @@ export default class EntityBuilder {
...
 
-    validateFieldsConstrainedToSet(params, allowedParametersList);
+    const castedParams = validateAndCast(params)
+      .fieldsConstrainedToSet(allowedParametersList)
+      .castNumber(['fromTimestamp', 'toTimestamp', 'page', 'perPage'])
+      .isNonNegativeInteger(['fromTimestamp', 'toTimestamp', 'page', 'perPage'])
+      .isAddress(['createdBy'])
+      .getCastedParams();
 
     this.ensureGeoLocationParamsCorrectlyPlaced(params);
```
Comment (Anze1m) [inline src/services/entity_builder.js]: Why isn't this also in `ValidationAndCasting `?

File: src/services/entity_builder.js [hunk 2]
```diff
@@ -41,47 +35,51 @@ export default class EntityBuilder {
...
+
     this.eventValidators.forEach((validator) => validator.validate(event.content));
-    this.ensureTimestampWithinLimit(event.content.idData.timestamp);
-    validateFieldsConstrainedToSet(event, ['content', 'eventId']);
-    validateFieldsConstrainedToSet(event.content, ['idData', 'data', 'signature']);
-    validateNonNegativeInteger(event.content.idData.accessLevel, `Access level should be a non-negative integer, instead got ${event.content.idData.accessLevel}`);
-    validateNonNegativeInteger(event.content.idData.timestamp, `Timestamp should be a non-negative integer, instead got ${event.content.idData.timestamp}`);
 
-    this.identityManager.validateSignature(event.content.idData.createdBy, event.content.signature, event.content.idData);
+    this.identityManager.validateSignature(event.content.idData.createdBy, event.content.signature,
```
Comment (Anze1m) [inline src/services/entity_builder.js]: Maybe do this validation via a custom validator in `ValidationAndCasting `?

Comment (vanruch) [reply src/services/entity_builder.js]: Checking signatures is much more complex then usual validations, I prefer having it separate

File: src/services/entity_builder.js (hunk 1 above)
Comment (vanruch) [reply src/services/entity_builder.js]: ValidationAndCasting is designed for simple checks of individual fields, checking geolocation is a bit more complex

File: src/services/entity_builder.js [hunk 3]
```diff
@@ -41,47 +35,51 @@ export default class EntityBuilder {
...
+        'assetId',
+        'content.idData',
+        'content.signature',
+        'content.idData.createdBy',
+        'content.idData.timestamp',
+        'content.idData.sequenceNumber'
+      ])
+      .fieldsConstrainedToSet(['content', 'assetId'])
+      .fieldsConstrainedToSet(['idData', 'signature'], 'content')
+      .validate(['content.idData.timestamp'], this.isTimestampWithinLimit.bind(this),
```
Comment (spherefoundry) [inline src/services/entity_builder.js]: I find `bind` particularly ugly, and would prefer an arrow function that calls `this.isTimestamp` instead.

**Pull request merged**